### PR TITLE
[CIRCLE-14594] Testing via local builds wrapped in orb tooling

### PR DESCRIPTION
**Pull request by cayennes**

### CODE REVIEW

**Comment by cayennes** [comment]
The output matching tooling isn't that specific - it simply allows you to specify steps that will be run while the output exists in a file, and you can use whatever tooling you want (grep, bats, etc.) to make assertions about that file.

Hopefully I'll get an example of that up soon.

**Comment by cayennes** [comment]
This is ready to be reviewed now. It needs more documentation but that is another step.

File: src/@orb.yml [hunk 1]
```diff
@@ -101,6 +101,35 @@ commands:
...
+      external-check-steps:
+        description: Checks you want to run that look at the contents of local_build_output.txt. Checks that depend on the environment inside the test job should go inside the test config. This stop should probably call a script rather than hardcode the checks into your config.
+        type: steps
+        default: []
+    steps:
+      - run: cp << parameters.test-config-location >> tmp-config-src/config.yml
+      - pack:
+          source: tmp-config-src
+          destination: uncompiled-config.yml
+      # run that config inside a local job - it will have pre-steps that contain the config to be packed inside it, and post-steps that check that it looks like it should
```
Comment (hannahhenderson) [inline src/@orb.yml]: ```run that config inside a local job``` to ```run uncompiled-config inside a local job``` for clarity

File: src/@orb.yml [hunk 2]
```diff
@@ -324,3 +353,49 @@ jobs:
...
           token-variable: << parameters.publish-token-variable >>
+
+  test-in-builds:
+    parameters:
+      orb-location:
+        type: string
+      orb-name:
+        type: string
+      checkout:
+        description: >
```
Comment (hannahhenderson) [inline src/@orb.yml]: This is cool. I didn't realize we could do this for a multi-line description

**Comment by eddiewebb** [comment]
General nit, I was under the impression we wanted to standardize on `snake_case` but mixed used of hyphens and underscores here.

I think functionality is solid, the question I have is around consumption.

`local-test-build` is mean to be an argument for `test-in-builds`'s `test_steps`, but itself includes `external-check-steps` so it feels odd to me that something that is itself meant to be a step argument takes step arguments, meaning the following configs should render identical output/

```
workflows:
  using_external_steps:
    jobs:
      orb-tools/test-in-builds:
        test_steps:
          - orb-tools/local-test-build:
              external-check-steps:
                - run: diff local_build_output.txt expected_output.txt
```

```
workflows:
  usings_test_steps:
    jobs:
      orb-tools/test-in-builds:
        test_steps:
          - orb-tools/local-test-build
          - run: diff local_build_output.txt expected_output.txt
```

1.  Is my understanding correct?
2. If there is a strong reason driving this design, I think we ought to have strong examples.

**Comment by cayennes** [comment]
there isn't a very strong reason for 2 - it felt better to me because it is dealing with state related to the local build, and it seemed more robust to not assume that this would always be like the current implementation where that lies around until the next one overwrites it.

**Comment by cayennes** [comment]
good catch on the snake vs. kebab - looks like that inconsistency predates this PR though, and kebab is more common than snake.

**Comment by cayennes** [comment]
after talking to @ndintenfass the conclusion was that kebab-case is fine - we've standardized on snake_case for primitives, but inside orbs it should just be consistent with itself. we're going to have an easier time being consistent with kebab-case. I checked and I don't see any snake_case for things defined by this orb.

**Pull request merged**